### DISCUSSION TITLE
Adding Support for Custom Parameter Injection into `QueryFunctionContext`

### DISCUSSION THREAD

**Discussion by kj455** [TanStack/query#6022] [Ideas] [answered]
### Feature Request
Adding a mechanism to support the Dependency Injection (DI) of arbitrary values into `QueryFunctionContext`.

### Background and Current Situation
As of now, the `queryFn` argument (`QueryFunctionContext`) supports only the following four elements:
- `queryKey`
- `pageParam`
- `signal`
- `meta`

I propose to extend this support to allow end-users to add arbitrary values, thereby facilitating a more flexible usage of the `QueryFunctionContext`.
### Detailed Proposal
My current idea is to enable this through the Dependency Injection (DI) of these values at the time of creating the `QueryClient`, as illustrated below:
```ts
const queryClient = new QueryClient({
  ...
  extraArgument: { /** inject something useful here */ } 
})
```
This mechanism would resemble the withExtraArgument function utilized in redux-thunk.
- [Reference to Redux Thunk's withExtraArgument](https://redux.js.org/usage/writing-logic-thunks#injecting-config-values-into-thunks))
### Motivation
Implementing the above proposal would facilitate easier writing of queryFn tests. Although queryFn generally involves API requests, which usually result in the following type of code:
```ts
// CURRENT
// queryFn
const fetchTodoList = () => {
  const { data } = api.get('/todos'); // api refers to any http client of your preference.
  return data
}

// usecase
const { data } = useQuery({
  queryKey: ['todo', 'list'],
  queryFn: fetchTodoList
})

// test
test('fetchTodoList', async () => {
  const api = { get: jest.fn() }
  jest.mock('path/to/api', () => api); // you have to mock http client...

  await fetchTodoList();

  expect(api.get).toHaveBeenCalledWith('/todos')
})
```
To test `fetchTodoList`, one needs to employ `jest.mock` to mock the HTTP client since it's used directly, making the tests prone to breaking. Hence, there is a desire to write tests using DI instead of module mocking.

If the proposed DI into `QueryFunctionContext` becomes feasible, it would substantially simplify testing:

```ts
// PROPOSED
// init
const queryClient = new QueryClient({
  extraArgument: { api } // inject http client
})

// queryFn
const fetchTodoList = ({ api }) => {
  const { data } = api.get('/todos'); // no need to api directly!!!
  return data
}

// usecase
const { data } = useQuery({
  queryKey: ['todo', 'list'],
  queryFn: fetchTodoList
})

// test
test('fetchTodoList', async () => {
  const api = { get: jest.fn() }

  await fetchTodoList({ api });  // no need to mock api!!

  expect(api.get).toHaveBeenCalledWith('/todos')
})
```
### Conclusion
I believe that the addition of this feature to the `QueryFunctionContext` would be highly beneficial. I would appreciate similar support in mutations as well. I am open to providing further details if needed, and if the technical complexity isn't high, I am motivated to contribute to its development.

I eagerly await your feedback and am open to discussing the technical feasibility and implications of this feature request further. Any suggestions or opinions on this feature request would be highly valued.

**Comment by TkDodo** [comment, marked as answer]
> I propose to extend this support to allow end-users to add arbitrary values, thereby facilitating a more flexible usage of the

I would say this is exactly what `meta` is for 😅 . From the docs:

> meta: Record<string, unknown>
Optional
If set, stores additional information on the query cache entry that can be used as needed. It will be accessible wherever the query is available, and is also part of the QueryFunctionContext provided to the queryFn.

```
const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      meta: {
       api // inject http client
     }
    }
  }
})
```

**Reply by kj455** [reply]
@TkDodo 
Understood, I overlooked the possibility of utilizing `meta` for that role 🙏 .
Thank you for pointing that out. It seems like a very straightforward solution.

On the other hand, could a similar solution involving placing values in `meta` work for mutations? 👀 
Given that the `mutationFn` argument only contains `variables`, I am wondering if finding a similar solution would be challenging. I apologize if this has already been covered in the docs and I missed it.

**Reply by TkDodo** [reply]
no, mutations don't have `meta`

**Reply by TkDodo** [reply]
generally, for what you want, my suggestion would rather be to create an `<ApiProvider>` and inject your api separately, and read from that in the queryFn as well as in the mutationFn.

**Reply by kj455** [reply]
Thank you for the swift advice. 
Indeed, creating an `<ApiProvider>` seems like a promising approach. Thank you for offering such a valuable perspective!

**Reply by kj455** [reply]
@TkDodo 
I apologize for asking another question.

I understand that injecting values such as `api` into `defaultOptions.queries.meta` would make them available in the `queryFn` arguments of all Queries.
However, if we set unique values in `meta` for individual Queries (like values for global error handling), it seems that the `meta` would be overridden, erasing the originally set values(e.g. `api`).

In this case, it seems feasible to solve this by obtaining the default `meta` using  `queryClient.getDefaultOptions` and merging it with the unique `meta` values, but I wonder if there are other alternatives?

```ts
const queryClient = useQueryClient();
const { data } = useQuery({
  queryFn: ({ meta: { api } }) => {...},
  meta: {
    ...queryClient.getDefaultOptions().meta,
    errorMessage: 'failed to fetch todos' // for global error
  }
})
```

If the above method is the most straightforward approach, it makes me feel that utilizing the suggested `<ApiProvider>` might be a more intuitive option.

**Reply by TkDodo** [reply]
yeah you'd need to merge like you've shown. I personally wouldn't set meta on a global level

**Reply by kj455** [reply]
I'm also leaning towards not setting the `meta` globally. Thank you!!
